compiler/rustc_codegen_ssa/src/lib.rs

```diff
@@ -298,7 +305,24 @@ impl CodegenResults {
         }
 
         let codegen_results = CodegenResults::decode(&mut decoder);
+        let lint_levels = CodegenLintLevels::decode(&mut decoder);
         let outputs = OutputFilenames::decode(&mut decoder);
-        Ok((codegen_results, outputs))
+        Ok((codegen_results, lint_levels, outputs))
+    }
+}
+
+/// A list of lint levels used in codegen.
+///
+/// When using `-Z link-only`, we don't have access to the tcx and must work
+/// solely from the `.rlink` file. `Lint`s are defined too early to be encodeable.
+/// Instead, encode exactly the information we need.
+#[derive(Copy, Clone, Encodable, Decodable)]
+pub struct CodegenLintLevels {
```

Comment:
Maybe put this in `CrateInfo`. That is already the place where we collect the various things that linking needs from the `TyCtxt`.

Reply:
oh that's a fantastic idea, that simplifies all the intermediate functions too so they don't have to pass along LintLevels explicitly, only the CrateInfo they were already passing